File: musinsa-api/src/main/java/com/moodfit/musinsa/service/MusinsaService.java
package com.moodfit.musinsa.service;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.moodfit.musinsa.dto.ImportResultDto;
import com.moodfit.musinsa.entity.MusinsaItem;
import com.moodfit.musinsa.repository.MusinsaItemRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.io.File;
import java.util.Iterator;
import java.util.Map;

@Service
@RequiredArgsConstructor
public class MusinsaService {

    private final MusinsaItemRepository itemRepository;
    private final ObjectMapper objectMapper = new ObjectMapper();

    public Page<MusinsaItem> findAll(int page, int size, String category) {
        PageRequest pageable = PageRequest.of(page, size, Sort.by("createdAt").descending());
        if (category != null && !category.isBlank()) {
            return itemRepository.findByCategory(category, pageable);
        }
        return itemRepository.findAll(pageable);
    }

    public MusinsaItem findById(String id) {
        return itemRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Item not found: " + id));
    }

    @Transactional
    public ImportResultDto importFromMetadata(String metadataPath) throws Exception {
        File file = new File(metadataPath);
        if (!file.exists()) {
            throw new RuntimeException("metadata.json not found: " + metadataPath);
        }

        JsonNode root = objectMapper.readTree(file);
        JsonNode categories = root.get("categories");

        int imported = 0;
        int skipped = 0;

        Iterator<Map.Entry<String, JsonNode>> catFields = categories.fields();
        while (catFields.hasNext()) {
            Map.Entry<String, JsonNode> catEntry = catFields.next();
            String category = catEntry.getKey();
            JsonNode items = catEntry.getValue().get("items");

            if (items == null || !items.isArray()) continue;

            for (JsonNode item : items) {
                String musinsaId = item.path("id").asText("");
                String imageUrl  = item.path("image_url").asText("");

                if (musinsaId.isBlank() || imageUrl.isBlank()) {
                    skipped++;
                    continue;
                }

                if (itemRepository.existsByMusinsaId(musinsaId)) {
                    skipped++;
                    continue;
                }

                MusinsaItem entity = new MusinsaItem();
                entity.setMusinsaId(musinsaId);
                entity.setName(item.path("name").asText("Unknown"));
                entity.setBrand(item.path("brand").asText(null));
                entity.setPrice(item.path("price").asInt(0));
                entity.setImageUrl(imageUrl);
                entity.setLocalPath(item.path("local_path").asText(null));
                entity.setCategory(category);

                itemRepository.save(entity);
                imported++;
            }
        }

        return new ImportResultDto(imported, skipped);
    }
}

File: apps/api/src/musinsa/musinsa.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { MusinsaItemEntity } from './entities/musinsa-item.entity';
import * as fs from 'fs';
import * as path from 'path';

@Injectable()
export class MusinsaService {
  constructor(
    @InjectRepository(MusinsaItemEntity)
    private readonly itemRepo: Repository<MusinsaItemEntity>,
  ) {}

  async findAll(options: {
    page?: number;
    limit?: number;
    category?: string;
  } = {}) {
    const { page = 1, limit = 24, category } = options;

    const qb = this.itemRepo.createQueryBuilder('item');

    if (category) {
      qb.where('item.category = :category', { category });
    }

    qb.orderBy('item.createdAt', 'DESC');

    const total = await qb.getCount();
    const data = await qb
      .skip((page - 1) * limit)
      .take(limit)
      .getMany();

    return {
      data,
      total,
      page,
      limit,
      totalPages: Math.ceil(total / limit),
    };
  }

  async findById(id: string): Promise<MusinsaItemEntity> {
    const item = await this.itemRepo.findOne({ where: { id } });
    if (!item) throw new NotFoundException('Musinsa item not found');
    return item;
  }

  async findByCategory(category: string): Promise<MusinsaItemEntity[]> {
    return this.itemRepo.find({ where: { category } });
  }

  /**
   * scripts/data/metadata.json 에서 크롤링 결과를 DB에 임포트합니다.
   * POST /musinsa/import 에서 호출됩니다.
   */
  async importFromMetadata(metadataPath?: string): Promise<{ imported: number; skipped: number }> {
    const filePath = metadataPath
      ? path.resolve(metadataPath)
      : path.resolve(process.cwd(), '../../scripts/data/metadata.json');

    if (!fs.existsSync(filePath)) {
      throw new NotFoundException(`metadata.json not found at: ${filePath}`);
    }

    const raw = fs.readFileSync(filePath, 'utf-8');
    const metadata = JSON.parse(raw);

    let imported = 0;
    let skipped = 0;

    for (const [category, catData] of Object.entries<any>(metadata.categories)) {
      for (const item of catData.items ?? []) {
        if (!item.id || !item.image_url) {
          skipped++;
          continue;
        }

        const existing = await this.itemRepo.findOne({ where: { musinsaId: item.id } });
        if (existing) {
          skipped++;
          continue;
        }

        await this.itemRepo.save(
          this.itemRepo.create({
            musinsaId: item.id,
            name: item.name ?? 'Unknown',
            brand: item.brand ?? null,
            price: item.price ?? 0,
            imageUrl: item.image_url,
            localPath: item.local_path ?? null,
            category,
          }),
        );
        imported++;
      }
    }

    return { imported, skipped };
  }
}
